File: 2022_11_03_leetcode_in_fsharp/LeetCode/LeetCode/P0392_IsSubsequence.fs
﻿// https://leetcode.com/problems/is-subsequence/

module LeetCode.P0392_IsSubsequence

let isSubsequence (s: string) (t: string) =
    s
    |> Seq.scan
        (fun i c ->
            let start = i + 1
            let len = t.Length - start
            { start .. start + len - 1 } |> Seq.tryFindIndex (fun j -> t[j] = c) |> Option.defaultValue -1)
        -1
    |> Seq.skip 1
    |> Seq.forall (fun i -> i <> -1)


let _ = isSubsequence "abc" "ahbgdc" // -> true
let _ = isSubsequence "axc" "ahbgdc" // -> false

File: 2022_11_19_leetcode_in_csharp/LeetCode/P0392_IsSubsequence.cs
// https://leetcode.com/problems/is-subsequence/

namespace LeetCode;

using static System.Linq.Enumerable;

class P0392_IsSubsequence
{
    static bool IsSubsequence(string s, string t)
    {
        return s
            .Scan(-1, (i, c) => Range(i + 1, t.Length - (i + 1)).FirstOrDefault(j => t[j] == c, -1))
            .Skip(1)
            .All(i => i != -1);
    }

    static bool IsSubsequence2(string s, string t)
    {
        return t
            .Scan(0, (i, c) => s[i] == c ? i + 1 : i)
            .Any(i => i == s.Length);
    }

    public static void Run()
    {
        var inputs = new[] {
            ("abc", "eopyahbgdc"),
            ("axc", "ahbgdc"),
        };

        foreach (var (s, t) in inputs)
        {
            var output = IsSubsequence(s, t);
            Console.WriteLine($" {(s, t)} -> {output}");
        }
    }
}